src/Lean/DocString/Links.lean

```diff
@@ -0,0 +1,152 @@
+is available, the latest version of the manual is used.
+-/
+def rewriteManualLinksCore (s : String) : BaseIO (Array (String.Range × String) × String) := do
+  let scheme := "lean-manual://"
+  let mut out := ""
+  let mut errors := #[]
+  let mut iter := s.iter
+  while h : iter.hasNext do
+    let c := iter.curr' h
+    iter := iter.next' h
+
+    if lookingAt scheme iter.prev then
+        let start := iter.prev.forward scheme.length
+        let mut iter' := start
+        while h' : iter'.hasNext do
+          let c' := iter'.curr' h'
+          iter' := iter'.next' h'
+          if !urlChar c' || iter'.atEnd then
+            match rw (start.extract iter'.prev) with
+            | .error err =>
+              errors := errors.push (⟨iter.prev.i, iter'.prev.i⟩, err)
+              out := out.push c
+              break
+            | .ok path =>
+              out := out ++ (← manualRoot) ++ path
+              out := out.push c'
+              iter := iter'
+              break
+    else
+      out := out.push c
+  pure (errors, out)
+
+where
+  /--
+  Returns `true` if the character is one of those allowed in RFC 3986 sections 2.2 and 2.3. other
+  than '(' or')'.
+  -/
+  urlChar (c : Char) : Bool :=
+    -- unreserved
+    c.isAlphanum || c == '-' || c == '.' || c == '_' || c == '~' ||
+    -- gen-delims
+    c == ':' || c == '/' || c == '?' || c == '#' || c == '[' || c == ']' || c == '@' ||
+    -- sub-delims
+    -- ( and ) are excluded due to Markdown's link syntax
+    c == '!' || c == '$' || c == '&' || c == '\'' || /- c == '(' || c == ')' || -/ c == '*' ||
+    c == '+' || c == ',' || c == ';' || c == '='
+
+  /--
+  Returns `true` if `goal` is a prefix of the string at the position pointed to by `iter`.
+  -/
+  lookingAt (goal : String) (iter : String.Iterator) : Bool := Id.run do
+    let mut iter := iter
+    let mut iter' := goal.iter
+    while h : (iter.hasNext ∧ iter'.hasNext) do
+      have ⟨h, h'⟩ := h
+      if iter.curr' h == iter'.curr' h' then
+        iter := iter.next' h
+        iter' := iter'.next' h'
+      else return false
+    return !iter'.hasNext
+
+  rw (path : String) : Except String String := do
+    match path.splitOn "/" with
+    | "section" :: args =>
+      if let [s] := args then
+        if s.isEmpty then
+          throw s!"Empty section ID"
+        return s!"find/?domain=Verso.Genre.Manual.section&name={s}"
+      else
+        throw s!"Expected one item after 'section', but got {args}"
+    | [] | [""] =>
+      throw "Missing documentation type"
+    | other :: _ =>
+      throw s!"Unknown documentation type '{other}'. Expected 'section'."
+
+
+/--
+Rewrites Lean reference manual links in `docstring` to point at the reference manual.
+
+This assumes that all such links have already been validated by `validateDocComment`.
+-/
+def rewriteManualLinks (docString : String) : BaseIO String := do
+  let (errs, str) ← rewriteManualLinksCore docString
+  assert! errs.isEmpty
```

Comment:
@mhuisi I wonder whether this line should instead append an error report to the docstring in question. The risk is that this might be missed in cases where docstrings are added via the low-level API by metaprogramming. On the other hand, it may be easier to recover from. I think I'll push a commit that does this, and we can revert it if you disagree.

Reply:
Updated in 9de9505. I think it's much better.